### ISSUE TITLE
Framerate problem with SharedTextureEnabled when NOT using ExternalBeginFrameEnabled

### ISSUE THREAD

**Issue by spellizzari** [cefsharp/CefSharp#2675]
I'm trying to implement something similar to cef-mixer but with CefSharp.

The context:
- Windows 10.1
- x64 build

**What works: using `ExternalBeginFrameEnabled`**
- with NuGet CefSharp.OffScreen v71.0.0. 
- CefSettings initialized with `CachePath` set, and the following `CefCommandLineArgs`: `disable-gpu-shader-disk-cache`, `show-fps-counter`, `use-angle=d3d11`.
- `ChromiumWebBrowser` created on url `https://www.google.com/` and `BrowserSettings` with `WindowlessFrameRate = 60`.
- a custom `IRenderHandler` implementation that implements `OnAcceleratedPaint` and performs D3D11/WPF interop through a `D3DImage` object (WPF specific class).
- I call `CreateBrowser` manually with a `WindowInfo` setup with `SetAsWindowless()`, and with `SharedTextureEnabled`, `WindowlessRenderingEnabled` **and** `ExternalBeginFrameEnabled` set to true.
- I subscribe to the `CompositionTarget.Rendering` WPF event to periodically call `SendExternalBeginFrame`.

I see the expected result: my image in my WPF window gets refreshed at 60fps (which is the expected rate as it's the rate at which WPF renders its composition) and the caret blinks in the search bar inside the Google page at the correct rate.

**What doesn't work: NOT using `ExternalBeginFrameEnabled`**

I do everything like the above, except:
- I don't set `ExternalBeginFrameEnabled` to true.
- I don't call `SendExternalBeginFrame`.

What I get is that my image in my WPF window gets refreshed at 1fps (it's the rate at which `OnAcceleratedPaint` gets called by CefSharp), and as a consequence the caret inside the search bar of the Google page blinks slowly.

**Attempts with cef-mixer and recompiling CefSharp**

I did some tests with cef-mixer (commit 1e941c9) by setting `window_info.external_begin_frame_enabled` to `false` in `web_layer.cpp:1168`. What I get is that after the page gets loaded, it gets refreshed at 5fps, which is the frequency of the blink animation of the caret inside the search bar.

I thought that this difference in behavior may be caused by the fact that CefSharp 71.0 and this version of cef-mixer don't rely on the same version of CEF: cef-mixer uses a patched version of cef 3.3599 (commit g285dbb1). So what I did is:
- cloned CefSharp commit 9c5f9c2,
- made manual versions of cef.redist.x64.* and cef.sdk.* package folders by using the same modified version of CEF as cef-mixer,
- successfully compiled a custom version of CefSharp

But when I use this custom version of CefSharp, I get exactly the same behavior than with NuGet version 71.0 of CefSharp: `OnAcceleratedPaint` is called at 1fps no matter what happens on the page.

I have no idea what can cause this behavior. I did those tests with cef-mixer prior to posting an issue here because I thought at first that this may be a CEF-related issue, but I'm starting to think it may be related to CefSharp. 

Any idea?

**Comment by spellizzari** [comment]
More news on the matter.

If I subscribe to the `BrowserInitialized` event on my browser instance, and call `_browser.GetBrowserHost().WindowlessFrameRate = 60;` inside it, I get the expected behavior (5fps) when using my custom CefSharp build (based on the same patched build of CEF that cef-mixer uses). The same trick doesn't change anything (still 1fps) if using NuGet version 71.0 or version 72.0.0-CI2978.

So if I sum it up:
- one problem seems related to CefSharp, which is that the value set on `WindowlessFrameRate` inside `BrowserSettings` seems to get replaced with value `1` when `SharedTextureEnabled` is set to true.
- one problem is CEF-related, which comes from how `OnAcceleratedPaint` has been implemented in CEF 71 and 72 which works worse than how cef-mixer did it in its own patched version.

**Comment by spellizzari** [comment]
More news on the matter.

If I subscribe to the `BrowserInitialized` event on my browser instance, and call `_browser.GetBrowserHost().WindowlessFrameRate = 60;` inside it, I get the expected behavior (5fps) when NuGet version 71.0.

So it seems there is a problem related to CefSharp, which is that the value set on `WindowlessFrameRate` inside `BrowserSettings` seems to get replaced with value `1` when `SharedTextureEnabled` is set to true. Maybe there's an unexpected BOOL to INT conversion somewhere in this codepath inside CefSharp that causes this particular value (1) to be used?

**Comment by amaitland** [comment]
> So it seems there is a problem related to CefSharp, which is that the value set on `WindowlessFrameRate` inside `BrowserSettings` seems to get replaced with value `1` when `SharedTextureEnabled` is set to true. Maybe there's an unexpected BOOL to INT conversion somewhere in this codepath inside CefSharp that causes this particular value (1) to be used?

Seems very unlikely. You can test with the `CefSharp.Wpf.Example` example (unmodified).

https://github.com/cefsharp/CefSharp/blob/master/CefSharp.Core/BrowserSettings.h#L374

Without a `PR` that actually implements `OnAcceleratedPaint ` I cannot help you as I have no means of reproducing this.

**Comment by spellizzari** [comment]
The problem occurs even if the OnAcceleratedPaint does nothing other than calculating its frame rate. It's possible to reproduce it with `CefSharp.MinimalExample.OffScreen`, I attached a modified version of it that reproduces the problem with https://google.com/:
- as is, the browser is initialized with offscreen rendering and a custom `IRenderHandler` that calculates a frame rate in `OnPaint`. After the page is loaded, the frame rate is constant and equal to 5, which is the expected behavior.
- if you uncomment line 1 `#define USE_SHARED_TEXTURE`, shared textures are used and the custom `IRenderHandler` calculates a frame rate in `OnAcceleratedPaint`. After the page is loaded, the frame rate is constant and equal to 1.
- if also you uncomment line 2 `#define RESET_WINDOWLESSFRAMERATE_ON_LOAD`, shared textures are used and `WindowlessFrameRate` is manually re-set to 60 after the browser is initialized. After the page is loaded, the frame rate is constant and equal to 5.

[CefSharp.MinimalExample.OffScreen.zip](https://github.com/cefsharp/CefSharp/files/2935040/CefSharp.MinimalExample.OffScreen.zip)

**Comment by amaitland** [comment]
> Seems very unlikely. You can test with the `CefSharp.Wpf.Example` example (unmodified).

Did you test with just setting the frame rate and enabling the frame rate counter with one of the `WPF` examples? 

> Please no binary (zip, etc) links, fork the MinimalExample and push your changes to GitHub. (Alternatively use a code sharing service list Gist or Pastebin).

Can you please push your code to `GitHub`, I cannot easily open a `zip` file from an `Android` device.

**Comment by amaitland** [comment]
A very quick test and that's all I have time for at the moment and setting `WindowlessFrameRate` appears to work as expected using the `show-fps-counter` command line arg. 

```c#
using System.Windows;

namespace CefSharp.MinimalExample.Wpf
{
    public partial class MainWindow : Window
    {
        public MainWindow()
        {
            InitializeComponent();

            Browser.BrowserSettings.WindowlessFrameRate = 15;
        }
    }
}
```

If/when someone contributes an `OnAcceleratedPaint ` implementation for the `WPF` project I'll see if I can have a look at this further.

In the meantime if you find the cause then by all means submit a `PR`.

**Comment by zhangkaixuan** [comment]
> I'm trying to implement something similar to cef-mixer but with CefSharp.
> 
> The context:
> 
> * Windows 10.1
> * x64 build
> 
> **What works: using `ExternalBeginFrameEnabled`**
> 
> * with NuGet CefSharp.OffScreen v71.0.0.
> * CefSettings initialized with `CachePath` set, and the following `CefCommandLineArgs`: `disable-gpu-shader-disk-cache`, `show-fps-counter`, `use-angle=d3d11`.
> * `ChromiumWebBrowser` created on url `https://www.google.com/` and `BrowserSettings` with `WindowlessFrameRate = 60`.
> * a custom `IRenderHandler` implementation that implements `OnAcceleratedPaint` and performs D3D11/WPF interop through a `D3DImage` object (WPF specific class).
> * I call `CreateBrowser` manually with a `WindowInfo` setup with `SetAsWindowless()`, and with `SharedTextureEnabled`, `WindowlessRenderingEnabled` **and** `ExternalBeginFrameEnabled` set to true.
> * I subscribe to the `CompositionTarget.Rendering` WPF event to periodically call `SendExternalBeginFrame`.
> 
> I see the expected result: my image in my WPF window gets refreshed at 60fps (which is the expected rate as it's the rate at which WPF renders its composition) and the caret blinks in the search bar inside the Google page at the correct rate.
> 
> **What doesn't work: NOT using `ExternalBeginFrameEnabled`**
> 
> I do everything like the above, except:
> 
> * I don't set `ExternalBeginFrameEnabled` to true.
> * I don't call `SendExternalBeginFrame`.
> 
> What I get is that my image in my WPF window gets refreshed at 1fps (it's the rate at which `OnAcceleratedPaint` gets called by CefSharp), and as a consequence the caret inside the search bar of the Google page blinks slowly.
> 
> **Attempts with cef-mixer and recompiling CefSharp**
> 
> I did some tests with cef-mixer (commit 1e941c9) by setting `window_info.external_begin_frame_enabled` to `false` in `web_layer.cpp:1168`. What I get is that after the page gets loaded, it gets refreshed at 5fps, which is the frequency of the blink animation of the caret inside the search bar.
> 
> I thought that this difference in behavior may be caused by the fact that CefSharp 71.0 and this version of cef-mixer don't rely on the same version of CEF: cef-mixer uses a patched version of cef 3.3599 (commit g285dbb1). So what I did is:
> 
> * cloned CefSharp commit [9c5f9c2](https://github.com/cefsharp/CefSharp/commit/9c5f9c285df664e695da823d07e75b2c611daa09),
> * made manual versions of cef.redist.x64.* and cef.sdk.* package folders by using the same modified version of CEF as cef-mixer,
> * successfully compiled a custom version of CefSharp
> 
> But when I use this custom version of CefSharp, I get exactly the same behavior than with NuGet version 71.0 of CefSharp: `OnAcceleratedPaint` is called at 1fps no matter what happens on the page.
> 
> I have no idea what can cause this behavior. I did those tests with cef-mixer prior to posting an issue here because I thought at first that this may be a CEF-related issue, but I'm starting to think it may be related to CefSharp.
> 
> Any idea?

hello, can you share your demo ?  i follow your steps， but， it does not work...

**Comment by amaitland** [comment]
The `SharedTexture` support has been temporarily removed as mentioned in https://bitbucket.org/chromiumembedded/cef/issues/2575/viz-implementation-for-osr#comment-53013860

There is an open [PR 285](https://bitbucket.org/chromiumembedded/cef/pull-requests/285/reimplement-shared-texture-support-for-viz/diff) though there's no signs it will be merged anytime soon.

If this problem can be reproduced after support is re-added then this can be reopened.
